esphome/components/sen5x/sen5x.cpp

```diff
@@ -331,79 +524,216 @@ void SEN5XComponent::update() {
       });
     }
   }
+  uint16_t cmd;
+  uint8_t length;
+  switch (this->model_.value()) {
+    case SEN60:
+      cmd = SEN60_CMD_READ_MEASUREMENT;
+      length = 4;
+      break;
+    case SEN63C:
+      cmd = SEN63_CMD_READ_MEASUREMENT;
+      length = 7;
+      break;
+    case SEN65:
+      cmd = SEN65_CMD_READ_MEASUREMENT;
+      length = 8;
+      break;
+    case SEN66:
+      cmd = SEN66_CMD_READ_MEASUREMENT;
+      length = 9;
+      break;
+    case SEN68:
+      cmd = SEN68_CMD_READ_MEASUREMENT;
+      length = 9;
+      break;
+    default:
+      cmd = SEN5X_CMD_READ_MEASUREMENT;
+      length = 9;
+      break;
+  }
 
-  if (!this->write_command(SEN5X_CMD_READ_MEASUREMENT)) {
+  if (!this->write_command(cmd)) {
     this->status_set_warning();
-    ESP_LOGD(TAG, "write error read measurement (%d)", this->last_error_);
+    ESP_LOGD(TAG, "Failed to write Measurement Command (last_error=%d)", this->last_error_);
     return;
   }
-  this->set_timeout(20, [this]() {
-    uint16_t measurements[8];
 
-    if (!this->read_data(measurements, 8)) {
+  this->set_timeout(20, [this, length]() {
+    uint16_t measurements[9];
+
+    if (!this->read_data(measurements, length)) {
       this->status_set_warning();
       ESP_LOGD(TAG, "read data error (%d)", this->last_error_);
       return;
     }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_1_0 = 0x%.4x", measurements[0]);
     float pm_1_0 = measurements[0] == UINT16_MAX ? NAN : measurements[0] / 10.0f;
+    if (this->pm_1_0_sensor_ != nullptr) {
+      this->pm_1_0_sensor_->publish_state(pm_1_0);
+    }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_2_5 = 0x%.4x", measurements[1]);
     float pm_2_5 = measurements[1] == UINT16_MAX ? NAN : measurements[1] / 10.0f;
+    if (this->pm_2_5_sensor_ != nullptr) {
+      this->pm_2_5_sensor_->publish_state(pm_2_5);
+    }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_4_0 = 0x%.4x", measurements[2]);
     float pm_4_0 = measurements[2] == UINT16_MAX ? NAN : measurements[2] / 10.0f;
+    if (this->pm_4_0_sensor_ != nullptr) {
+      this->pm_4_0_sensor_->publish_state(pm_4_0);
+    }
```

Comment:
There is no point doing maths that wont be used
```suggestion
    if (this->pm_4_0_sensor_ != nullptr) {
      float pm_4_0 = measurements[2] == UINT16_MAX ? NAN : measurements[2] / 10.0f;
      this->pm_4_0_sensor_->publish_state(pm_4_0);
    }
```

Reply:
Reworked all of these so that logging and maths are only executed if the sensor exists.